File: src/test/java/cilly/MembersTest.java
package cilly;

import java.io.PrintStream;

public class MembersTest {

    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("usage: java test.MembersTest assembly [classname]");
            System.exit(1);
        }

        Assembly mscorlib = Assembly.loadFrom("/usr/lib/mono/2.0/mscorlib.dll");
        Type.initMSCORLIB(mscorlib);
        Assembly assem = Assembly.loadFrom(args[0]);
        if (args.length > 1) {
            Type type = assem.getType(args[1]);
            if (type != null)
                dumpMember(System.out, type);
            else
                System.err.println("Cannot find type " + args[1] + " in "
                        + assem);
        } else {
            Type[] types = assem.getTypes();
            System.out.println("Number of types in assembly " + assem + " -> "
                    + types.length);
            dumpCustomAttributes(System.out, "assembly: ", assem);
            Module[] modules = assem.getModules();
            for (int i = 0; i < modules.length; i++) {
                dumpCustomAttributes(System.out, "module " + modules[i] + ": ",
                        modules[i]);
            }
            dumpMembers(System.out, types);
        }
    }

    public static final void dumpMember(PrintStream out, MemberInfo member) {
        try {
            if (member.memberType() == MemberTypes.TypeInfo
                    || member.memberType() == MemberTypes.NestedType) {
                Type type = (Type) member;
                dumpCustomAttributes(out, "", type);
                out.print(TypeAttributes.accessModsToString(type.Attributes));
                out.print(type.isInterface() ? " interface " : " class ");
                out.print(type);
                if (type.BaseType() != null)
                    out.println(" extends " + type.BaseType());
                Type[] ifaces = type.getInterfaces();
                if (ifaces.length > 0) {
                    out.print("\timplements ");
                    for (int i = 0; i < ifaces.length; i++) {
                        out.print(ifaces[i]);
                        if (i < (ifaces.length - 1))
                            out.print(", ");
                    }
                    out.println();
                }
                out.println("{");
                int all = BindingFlags.Public | BindingFlags.DeclaredOnly// |
                                                                         // BindingFlags.NonPublic
                        | BindingFlags.Instance | BindingFlags.Static;
                dumpMembers(out, type.getNestedTypes());
                dumpMembers(out, type.getFields(all));
                dumpMembers(out, type.getConstructors(all));
                dumpMembers(out, type.getMethods(all));
                dumpMembers(out, type.getProperties(all));
                dumpMembers(out, type.getEvents());
                out.println("}");
            } else {
                dumpCustomAttributes(out, "", member);
                out.print(MemberTypes.toString(member.memberType()));
                out.print(": ");
                out.print(member);
                out.println();
            }
        } catch (Throwable e) {
            String message = MemberTypes.toString(member.memberType()) + ": "
                    + member;
            throw new RuntimeException(message, e);
        }
    }

    public static void dumpCustomAttributes(PrintStream out, String prefix,
            ICustomAttributeProvider att) {
        Object[] attrs = att.getCustomAttributes(false);
        for (int j = 0; j < attrs.length; j++)
            out.println(prefix + attrs[j]);
    }

    public static void dumpMembers(PrintStream out, MemberInfo[] members) {
        for (int i = 0; i < members.length; i++) {
            dumpMember(out, members[i]);
        }
    }

}

File: src/test/scala/cilly/MembersTest.scala
package cilly

import java.io.PrintStream

object MembersTest {
  def main(args: Array[String]): Unit = {
    if (args.length < 1) {
      System.err.println("usage: java test.MembersTest assembly [classname]")
      System.exit(1)
    }
    val mscorlib: Assembly = Assembly.loadFrom("/usr/lib/mono/2.0/mscorlib.dll")
    Type.initMSCORLIB(mscorlib)
    val assem: Assembly = Assembly.loadFrom(args(0))
    if (args.length > 1) {
      val `type`: Type = assem.getType(args(1))
      if (`type` != null) dumpMember(System.out, `type`)
      else System.err.println("Cannot find type " + args(1) + " in " + assem)
    }
    else {
      val types: Array[Type] = assem.getTypes
      System.out.println("Number of types in assembly " + assem + " -> " + types.length)
      dumpCustomAttributes(System.out, "assembly: ", assem)
      val modules: Array[Module] = assem.modules
      
      {
        var i: Int = 0
        while (i < modules.length) {
          {
            dumpCustomAttributes(System.out, "module " + modules(i) + ": ", modules(i))
          }
          ({
            i += 1; i - 1
          })
        }
      }
      dumpMembers(System.out, types)
    }
  }

  def dumpMember(out: PrintStream, member: MemberInfo): Unit = {
    try {
      if (member.memberType == MemberTypes.TypeInfo || member.memberType == MemberTypes.NestedType) {
        val typ: Type = member.asInstanceOf[Type]
        dumpCustomAttributes(out, "", typ)
        out.print(TypeAttributes.accessModsToString(typ.attributes))
        out.print(if (typ.isInterface) " interface " else " class ")
        out.print(typ)
        if (typ.BaseType != null) out.println(" extends " + typ.BaseType)
        val ifaces: Array[Type] = typ.getInterfaces
        if (ifaces.length > 0) {
          out.print("\timplements ")
          
          {
            var i: Int = 0
            while (i < ifaces.length) {
              {
                out.print(ifaces(i))
                if (i < (ifaces.length - 1)) out.print(", ")
              }
              ({
                i += 1; i - 1
              })
            }
          }
          out.println()
        }
        out.println("{")
        val all: Int = BindingFlags.Public | BindingFlags.DeclaredOnly | BindingFlags.Instance | BindingFlags.Static
        dumpMembers(out, typ.getNestedTypes)
        dumpMembers(out, typ.getFields(all))
        dumpMembers(out, typ.getConstructors(all))
        dumpMembers(out, typ.getMethods(all))
        dumpMembers(out, typ.getProperties(all))
        dumpMembers(out, typ.getEvents)
        out.println("}")
      }
      else {
        dumpCustomAttributes(out, "", member)
        out.print(MemberTypes.toString(member.memberType))
        out.print(": ")
        out.print(member)
        out.println()
      }
    }
    catch {
      case e: Throwable =>
        val message: String = MemberTypes.toString(member.memberType) + ": " + member
        throw new RuntimeException(message, e)
    }
  }

  def dumpCustomAttributes(out: PrintStream, prefix: String, att: ICustomAttributeProvider): Unit = {
    val attrs: Array[AnyRef] = att.getCustomAttributes(inherit = false)
    
    {
      var j: Int = 0
      while (j < attrs.length) {
        out.println(prefix + attrs(j))
        j += 1
      }
    }
  }

  def dumpMembers(out: PrintStream, members: Array[_ <: MemberInfo]): Unit = {
    {
      var i: Int = 0
      while (i < members.length) {
        {
          dumpMember(out, members(i))
        }
        ({
          i += 1; i - 1
        })
      }
    }
  }
}